http_parser.c

```diff
@@ -1782,9 +1782,9 @@ size_t http_parser_execute (http_parser *parser,
 
         if (parser->flags & F_TRAILING) {
           /* End of a chunked request */
-          UPDATE_STATE(NEW_MESSAGE());
-          CALLBACK_NOTIFY(message_complete);
-          break;
+          UPDATE_STATE(s_message_done);
```

Comment:
The previous code didn't call the chunk_complete at this location at all.  Setting the state to s_message_done before calling the chunk_complete callback is necessary for correctness in case the chunk_complete pauses the parser before returning.  This ensures that the state will be s_message_done on the next execution even if CALLBACK_NOTIFY_NOADVANCE(chunk_complete) ends up returning.  This makes sure that the message_complete callback will still be called correctly after the chunk_complete callback.

If you are asking for the overall motivation for adding these chunking callback, this helps implement proxies more efficiently.  Even though the chunk bounadaries shouldn't have semantic meaning, it is more efficient for the proxy to retransmit the message using the exact same chunk boundaries as originally used in the message.  Otherwise it is pessimistically forced to rechunk with smaller chunk boundaries in some cases, resulting in more data transmitted on the wire (due to more chunk headers).